### ISSUE TITLE
Performance degradation in singlenode on upgrade from 1.85.3 to 1.88.1/1.89.1

### ISSUE THREAD

**Issue by Alexul** [VictoriaMetrics/VictoriaMetrics#3966]
### Describe the bug

After upgrade single node tsdb from 1.85.3 to 1.88.1 (and 1.89.1 doesn't help, same issue) hit a big performance degradation in both query latency and CPU usage
[metrics snapshot](https://grafana.wixpress.com/dashboard/snapshot/aaItcXCwWaRkGSc0zSu0jL43u7lfClx4)
<img width="1362" alt="Screenshot 2023-03-16 at 12 56 02" src="https://user-images.githubusercontent.com/10867201/225596306-2faacdb9-2329-474d-8bb0-f8f0fcd5ff3a.png">

### To Reproduce

upgrade to latest

### Version

victoria-metrics-20230228-015149-tags-v1.88.1-0-g95e117342
and latest release of 1.89.1 later

**Comment by Alexul** [comment]
dumps from the latest 1.89.1
[cpu.pprof.txt](https://github.com/VictoriaMetrics/VictoriaMetrics/files/10989804/cpu.pprof.txt)
[mem.pprof.txt](https://github.com/VictoriaMetrics/VictoriaMetrics/files/10989805/mem.pprof.txt)

**Comment by zekker6** [comment]
Hello @Alexul, thanks for sharing profiles!

This issue is very likely to be caused by updates in https://github.com/VictoriaMetrics/VictoriaMetrics/commit/53d871d0b173242c8f4ce13defd81abb861c358a . This commit changed parallel parsing in order to speed up overall processing speed for the cases when some CPU cores could finish their jobs earlier than others and would just wait for parsing completion. Change added work "stealing" mechanism so that workers that finished parsing of their own data blocks would try to "steal" remaining work from other workers. This involves calls to a global lock and can cause high CPU usage by locks synchronization between different CPUs on systems with higher number of CPUs.

We'll reproduce this case and prepare a patch to lower "stealing" rate in order to avoid using global locks, that should decrease CPU usage in general. 
Once this will be ready I'll share an image to test and verify that it helps for your case.

**Comment by zekker6** [comment]
@Alexul Could you please also share mutex profile for this workload?
It can be obtained by using `debug/pprof/mutex` endpoint.

**Comment by Alexul** [comment]
Sure, heres from the victoria-metrics-20230313-022110-tags-v1.89.1-0-g388d6ee16
[dashboard](https://grafana.wixpress.com/dashboard/snapshot/jCjk6Iuw5MHEOJHmsZRU61PGWV27r410)
[cpu.pprof.txt](https://github.com/VictoriaMetrics/VictoriaMetrics/files/11007934/cpu.pprof.txt)
[mutex.pprof.txt](https://github.com/VictoriaMetrics/VictoriaMetrics/files/11007935/mutex.pprof.txt)
[mem.pprof.txt](https://github.com/VictoriaMetrics/VictoriaMetrics/files/11007936/mem.pprof.txt)

**Comment by zekker6** [comment]
@Alexul Here is a built from https://github.com/VictoriaMetrics/VictoriaMetrics/tree/fix/3966-multi-cpu-optimization branch with several optimizations to mitigate mentioned issues:
```
docker.io/victoriametrics/victoria-metrics:heads-fix-3966-multi-cpu-optimization-0-g3d24acc53
```
Could you please run some testing and share new profiles so that we can take a look at optimizations results for this case?

**Comment by Alexul** [comment]
Much better, but still slower then 1.85.3
<img width="1357" alt="Screenshot 2023-03-23 at 06 53 14" src="https://user-images.githubusercontent.com/10867201/227106914-1da74939-b279-48db-bb51-a5dc6eb48a69.png">
[cpu.pprof.txt](https://github.com/VictoriaMetrics/VictoriaMetrics/files/11047082/cpu.pprof.txt)
[mutex.pprof.txt](https://github.com/VictoriaMetrics/VictoriaMetrics/files/11047083/mutex.pprof.txt)
[mem.pprof.txt](https://github.com/VictoriaMetrics/VictoriaMetrics/files/11047084/mem.pprof.txt)

**Comment by zekker6** [comment]
@Alexul Could you please test this image and share profiles?
 ```
docker.io/victoriametrics/victoria-metrics:heads-fix-3966-multi-cpu-optimization-0-g7c3b541da
```

**Comment by valyala** [comment]
@Alexul , thanks for the assistance on narrowing down and fixing the performance degradation in VictoriaMetrics! All the improvements from @zekker6 mentioned above have been integrated into [VictoriaMetrics v1.87.4 LTS release](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/tag/v1.87.4). Could you install it and verify whether it shows the performance comparable or better than v1.85.3 on your workload?

**Comment by Alexul** [comment]
thank you for a quick fix, confirming comparable performance with v1.85.3 on that workload

**Comment by valyala** [comment]
@Alexul , could you send CPU, memory and block profiles collected from v1.87.4 ? This may help identify and remove the remaining scalability bottlenecks for systems with many CPU cores.

**Comment by Alexul** [comment]
Sure, [dashboard snapshot](https://grafana.wixpress.com/dashboard/snapshot/10jJOIJlP3p6IBmH8Q9bjosjhGfckmIY)
[cpu.pprof.txt](https://github.com/VictoriaMetrics/VictoriaMetrics/files/11086246/cpu.pprof.txt)
[mutex.pprof.txt](https://github.com/VictoriaMetrics/VictoriaMetrics/files/11086247/mutex.pprof.txt)
[mem.pprof.txt](https://github.com/VictoriaMetrics/VictoriaMetrics/files/11086250/mem.pprof.txt)

**Comment by valyala** [comment]
@Alexul , thanks for the provided profiles for v1.87.4 . They look good and we couldn't find low-hanging fruits on how to improve the scalability / performance / resource usage further.

All the optimizations mentioned above have been included in [VictoriaMetrics v1.90.0](https://github.com/VictoriaMetrics/VictoriaMetrics/releases). Closing the issue then as resolved.
